middleware/instrument.go

```diff
@@ -105,11 +105,15 @@ func (i Instrument) getRouteName(r *http.Request) string {
 func getRouteName(routeMatcher RouteMatcher, r *http.Request) string {
 	var routeMatch mux.RouteMatch
 	if routeMatcher != nil && routeMatcher.Match(r, &routeMatch) {
-		if name := routeMatch.Route.GetName(); name != "" {
-			return name
-		}
-		if tmpl, err := routeMatch.Route.GetPathTemplate(); err == nil {
-			return MakeLabelValue(tmpl)
+		if routeMatch.Route != nil {
+			if name := routeMatch.Route.GetName(); name != "" {
+				return name
+			}
+			if tmpl, err := routeMatch.Route.GetPathTemplate(); err == nil {
+				return MakeLabelValue(tmpl)
+			}
+		} else if routeMatch.MatchErr == mux.ErrNotFound {
+			return "notfound"
```

Comment:
This is getting pretty nested, lets straight line:

```suggestion
	var routeMatch mux.RouteMatch
	if routeMatcher == nil || !routeMatcher.Match(r, &routeMatch) {
		return ""
	}
	
	if routeMatch.MatchErr == mux.ErrNotFound {
		return "notfound"
	}
	
	if routeMatch.Route == nil {
 		return ""
	}
	
	if name := routeMatch.Route.GetName(); name != "" {
		return name
	}
	
	tmpl, err := routeMatch.Route.GetPathTemplate();
	if err == nil {
		return ""
	}
```

Reply:
Sure, done: https://github.com/weaveworks/common/pull/213/commits/005f255b268ac060819f0cdba998c2af000cbca1

(Didn't commit the suggestion as it was missing the `MakeLabelValue(tmpl)` in the last return 😌

Reply:
+1 yeah that code was just an example.